### PR TITLE
OpenAI refactoring

### PR DESCRIPTION
**Pull request by FlorianJoncour**
This is a reset of #2210.

The final goal is to implement function calls using the OpenAI API.
But since it was likely too much all at once, we will do it in two parts.

This pull request is only a refactoring/relocation of code to separate the Uvicorn server, the chat, and the completions.
The chat and completions are now in separate classes.
The goal is to make the entire codebase clearer and more easily modifiable in the future, as the completion should now be considered legacy.

The chat part has been divided into several methods, while the completion remained largely unchanged except for being encapsulated within a class.

Tested chat and completions with and without stream mode.

### CODE REVIEW

**Review by NikolaBorisov — APPROVED** [review]
I think this looks ok.

File: vllm/entrypoints/openai/serving_engine.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    async def _post_init(self):
+        engine_model_config = await self.engine.get_model_config()
+        self.max_model_len = engine_model_config.max_model_len
+
+        # A separate tokenizer to map token IDs to strings.
+        self.tokenizer = get_tokenizer(
+            engine_model_config.tokenizer,
+            tokenizer_mode=engine_model_config.tokenizer_mode,
+            trust_remote_code=engine_model_config.trust_remote_code)
+        self._load_chat_template(self.chat_template)
```
Comment (simon-mo) [inline vllm/entrypoints/openai/serving_engine.py:52]: chat template is the responsibility of ChatCompletion only

**Review by simon-mo — APPROVED** [review]
two minor points

**Comment by simon-mo** [comment]
will fix and merge this once #2355 is in.

**Comment by FlorianJoncour** [comment]
Fine.

I still made the changes ^^

**Comment by simon-mo** [comment]
@FlorianJoncour, merged! Thank you for the contribution, looking forward to the tool calling PR!

**Pull request merged**